ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/pull-queries-against-materialized-aggregates.json

```diff
@@ -2060,6 +2060,171 @@
+      "statements": [
+        "CREATE STREAM INPUT (ID INTEGER KEY, IGNORED INT) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE TABLE AGGREGATE AS SELECT ID, COUNT(1) AS COUNT FROM INPUT WINDOW TUMBLING(SIZE 1 SECOND) GROUP BY ID;",
+        "SELECT * FROM AGGREGATE WHERE ID + 10 = 20;",
+        "SELECT * FROM AGGREGATE WHERE ID + 10 = 20 and WINDOWSTART > 12500;",
+        "SELECT id, WINDOWSTART, WINDOWEND, COUNT, FLOOR(SQRT(WINDOWSTART)) as val FROM AGGREGATE WHERE SQRT(WINDOWSTART) > 110.0;",
+        "SELECT id, WINDOWSTART, WINDOWEND, COUNT, floor(exp(ID)) as val  FROM AGGREGATE WHERE floor(exp(ID)) > 30000;",
+        "SELECT id, WINDOWSTART, WINDOWEND, COUNT, floor(exp(ID)) as val  FROM AGGREGATE WHERE floor(exp(ID)) > 30000 AND WINDOWEND < 13500;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "timestamp": 12345, "key": 10, "value": {"val": 1}},
+        {"topic": "test_topic", "timestamp": 12346, "key": 11, "value": {"val": 2}},
+        {"topic": "test_topic", "timestamp": 13346, "key": 11, "value": {"val": 3}},
+        {"topic": "test_topic", "timestamp": 13345, "key": 10, "value": {"val": 4}}
+      ],
+      "responses": [
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"query": [
+          {"header":{"schema":"`ID` INTEGER KEY, `WINDOWSTART` BIGINT KEY, `WINDOWEND` BIGINT KEY, `COUNT` BIGINT"}},
+          {"row":{"columns":[10, 12000, 13000, 1]}},
+          {"row":{"columns":[10, 13000, 14000, 1]}}
+        ]},
+        {"query": [
+          {"header":{"schema":"`ID` INTEGER KEY, `WINDOWSTART` BIGINT KEY, `WINDOWEND` BIGINT KEY, `COUNT` BIGINT"}},
+          {"row":{"columns":[10, 13000, 14000, 1]}}
+        ]},
+        {"query": [
+          {"header":{"schema":"`ID` INTEGER KEY, `WINDOWSTART` BIGINT KEY, `WINDOWEND` BIGINT KEY, `COUNT` BIGINT, `VAL` DOUBLE"}},
+          {"row":{"columns":[10, 13000, 14000, 1, 114.0]}},
+          {"row":{"columns":[11, 13000, 14000, 1, 114.0]}}
+        ]},
+        {"query": [
+          {"header":{"schema":"`ID` INTEGER KEY, `WINDOWSTART` BIGINT KEY, `WINDOWEND` BIGINT KEY, `COUNT` BIGINT, `VAL` DOUBLE"}},
+          {"row":{"columns":[11, 12000, 13000, 1, 59874.0]}},
+          {"row":{"columns":[11, 13000, 14000, 1, 59874.0]}}
+        ]},
+        {"query": [
+          {"header":{"schema":"`ID` INTEGER KEY, `WINDOWSTART` BIGINT KEY, `WINDOWEND` BIGINT KEY, `COUNT` BIGINT, `VAL` DOUBLE"}},
+          {"row":{"columns":[11, 12000, 13000, 1, 59874.0]}}
+        ]}
+      ]
+    },
+    {
+      "name": "non-windowed - Table Scan - non column ref is not literal",
+      "statements": [
+        "CREATE STREAM INPUT (ID INTEGER KEY, IGNORED INT) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE TABLE AGGREGATE AS SELECT ID, COUNT(1) AS COUNT FROM INPUT GROUP BY ID;",
+        "SELECT * FROM AGGREGATE WHERE ID = 20 - 10;",
```

Comment:
```sql
SELECT * FROM AGGREGATE WHERE ID = 20 - 10;
```

at some point, this should result in a Key lookup right? What's the status of #4484 after we merge this PR? We might want to consider renaming that one (or opening a new one). I think it's important to consider that alongside this PR because we'd probably want a single-pass strategy (naively, we could just use the GenericExpressionResolver to try to resolve the non-key side of the equation and if it throws we assume it's not resolvable).

Reply:
Yes, this should result in a key lookup.  It should fall into the case of non column ref needs to be resolved.

I think #4484 should be resolved after this.

I was debating whether to do it that way, and I'm not opposed to doing it like that.  The current code is a common pattern where there's a validation and an extraction pass.  In this case, it makes dealing with the consequences a bit easier (we can just declare table scan required with the other similar checks and avoid extraction). Doing some of this in extraction would just slightly complicate that code a bit. Or did you mean merging Validation and Extraction more broadly to a single pass?

Reply:
Ah ignore that original comment, I hadn't properly understood the code on my first pass. Originally I had thought that we would table scan for `ID = 20 - 10` but now I'm seeing that we don't. I think the current implementation is fine.